### ISSUE TITLE
peigen.c:4081:1: internal compiler error: in rtx_equal_for_memref_p, at alias.c:1856 [ALL  ] 

### ISSUE THREAD

**Issue by 8ohmh** [crosstool-ng/crosstool-ng#867]
Hi!

Got a strange error:
```
[ALL  ]    peigen.c: In function 'rsrc_sort_entries.part.7':
[ERROR]    peigen.c:4081:1: internal compiler error: in rtx_equal_for_memref_p, at alias.c:1856
```
Crosstools-Ng `Version: 1.23.0`
Host OS: D`ebian 7.3 chrooted on a ARM Marvell CPU 
`
Build log, Config.log and more in file

**Remark**: I got fortunately the ARM Marvell Compiler working "compiled" on a i386 Debian 9.2.1 Version (Host: ARM Marvell, Target: ARM Marvell, Build: i386) and compiled with that new ARM Marvell Compiler on the ARM Board the whole compiler suite new for Target i386 (Host: ARM Marvell, Target: i386, Build: ARM Marvell), but now got those weird error

`e141501d29409abfdd580f1f71916998  peigen.tar.gz`

[peigen.tar.gz](https://github.com/crosstool-ng/crosstool-ng/files/1484799/peigen.tar.gz)

PS: Getting Mad- Seriously, only the i386 crosstools-ng environment is working right

**Comment by stilor** [comment]
First, why do you configure this as a canadian if your .config has BUILD and HOST machines both as arm-marvell-linux-gnueabi? This is a simple cross, doing it as "canadian" just adds the time and complexity.

The problem is, as the build log says, in the host compiler. Was it built with crosstool-NG? If so, attach the .config for the host toolchain.

Is it also a Linaro version or a stock GCC? You may need submit a bug upstream (i.e. to Linaro or to GCC).

**Comment by 8ohmh** [comment]
To1 .: Because other configs (cross-native, cross etc.) made the same (or other) unknown errors. As I wrote in the GMP issue I'm trying to get a working gcc arm marvell compiler for 3.2.40 Linux on that board - and after trying many parameters - that only with that canadian cross (i got)  so far that I almost got a working toolchain - but almost...

**Comment by stilor** [comment]
Well, as I said - the problem is the *host* compiler. If it is broken, it will break in the same or different ways on compiling complex code.

The question is, where does arm-marvell-linux-gnueabi-gcc come from? Is it from crosstool-NG? If so, attach **its** config.

**Comment by 8ohmh** [comment]
Thanks for the answers, but sorry, I just want to get a working toolchain - and a crosstools-ng which is working on different hosts correctly.. don't want to guess around about **your** tool

a28bd9384d25e9054c68c66543f1f8ff  ./arm-marvell-working.tar.gz

[arm-marvell-working.tar.gz](https://github.com/crosstool-ng/crosstool-ng/files/1484899/arm-marvell-working.tar.gz)

**Comment by 8ohmh** [comment]
Seems that's a GCC 6.3 ARM related error in generating in the parts of Portable Execute Files (???, didn't have enabled Mingw)

**Comment by 8ohmh** [comment]
I've tried also following:

x86-i386 -> ARM v3
ARM v3 .> ARM v9

Everything isn't working, trying now to compile on a VM

**Comment by 8ohmh** [comment]
Fault: Forgot to do a` ct-ng clean` before any new start of ct.ng and forgot to put in `CFlags` an option `"-mcpu=marvell-pj4"`

**Comment by 8ohmh** [comment]
After `ct-ng clean the` error now occurs in 

`Makefile:2733 all_build_libcpp Error 2`
function bool _parse_cpp_expr
libcpp.c, expr.c: 1345

Compiler: GCC 6.3.0

Question: Can I ever get a working toolchain by compiling on the ARM board as Host???

**Comment by stilor** [comment]
Answer: I don't know. It is likely that you're the first one to try it out.

You seem to miss the point: crosstool-NG just provides the scaffolding to build all the components in a single run, so any bugs should still be submitted to those individual projects. I don't have an ARM board to try this on, and building toolchain in a simulator will take forever. Hence it befalls on you to investigate/submit bugreport to GCC/discuss and test patches/etc.

In other words, I am not of much help to you here - not because I don't want to, but because I cannot.

**Comment by 8ohmh** [comment]
Thansk 4 the answer. Ok, how should I sent them the bug - all files in .build (the config etc. log of course, but I think to sent them 200 MB of files wouldn't be good ... but the bug in the file is in relation to other files...) And what about the GCC version 7.2 in the repo? Before I start again to write post about bugs that perhaps already fixed...

**Comment by stilor** [comment]
It is working on many configs, but the typical build environment is x86 (mostly linux, but osx/freebsd/cygwin are more or less supported too).

Typically, this includes the preprocessed source on which GCC is failing. Run the same command line, but replace `-c -o foo.o` with `-E -o foo.preproc.c` + the configure string for gcc (you can find it in the config log). You should also remove as many patches as possible when building the *arm* version of the toolchain (you can configure a "local" patch directory instead of "bundled" in crosstool-ng, copy the `patches` directory from the crosstool-ng installation into that local directory; then remove gcc patches (one by one, or binary search) until only the patches that are essential to getting gcc to compile remain. This is to exclude any backported patches.

GCC 7.2 is on crosstool-ng master branch - you can check out from Git and run the same procedure (but note that `patches` directory is now called `packages` in master).

**Comment by stilor** [comment]
Seems like an upstream bug.
